pkg/controllers/csi/driver/volumes/app/publisher.go

```diff
@@ -22,163 +22,166 @@ import (
+			"version or digest is not yet set, csi-provisioner hasn't finished setup yet for DynaKube: "+volumeCfg.DynakubeName,
 		)
 	}
 
-	if err := publisher.ensureMountSteps(ctx, bindCfg, volumeCfg); err != nil {
-		return nil, err
+	if err := pub.mountOneAgent(volumeCfg); err != nil {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("failed to mount oneagent volume: %s", err))
 	}
 
-	agentsVersionsMetric.WithLabelValues(bindCfg.MetricVersionLabel()).Inc()
-
 	return &csi.NodePublishVolumeResponse{}, nil
 }
 
-func (publisher *AppVolumePublisher) UnpublishVolume(ctx context.Context, volumeInfo *csivolumes.VolumeInfo) (*csi.NodeUnpublishVolumeResponse, error) {
-	volume, err := publisher.loadVolume(ctx, volumeInfo.VolumeID)
-	if err != nil {
-		log.Info("failed to load volume info", "error", err.Error())
-	}
+// hasRetryLimitReached creates the base dir for a given app mount if it doesn't exist yet, checks the creation timestamp against the threshold
+// if any of the FS calls fail in an unexpected way, then it is considered that the limit was reached.
+func (pub *Publisher) hasRetryLimitReached(volumeCfg *csivolumes.VolumeConfig) bool {
+	appDir := pub.path.AppMountForID(volumeCfg.VolumeID)
 
-	if volume == nil {
-		return &csi.NodeUnpublishVolumeResponse{}, nil
-	}
+	stat, err := pub.fs.Stat(appDir)
+	if errors.Is(err, os.ErrNotExist) {
+		// First run, create folder, to keep track of time
+		err := pub.fs.MkdirAll(appDir, os.ModePerm)
+		if err != nil {
+			log.Error(err, "failed to create base dir for app mount, skipping injection", "dir", appDir)
 
-	log.Info("loaded volume info", "id", volume.VolumeID, "pod name", volume.PodName, "version", volume.Version, "dynakube", volume.TenantUUID)
+			return true
+		}
 
-	if volume.Version == "" {
-		log.Info("requester has a dummy volume, no node-level unmount is needed")
+		return false
+	} else if err != nil {
+		log.Error(err, "unexpected failure in checking filesystem state, skipping injection", "dir", appDir)
 
-		return &csi.NodeUnpublishVolumeResponse{}, publisher.db.DeleteVolume(ctx, volume.VolumeID)
+		return true
 	}
 
-	overlayFSPath := publisher.path.AgentRunDirForVolume(volume.TenantUUID, volumeInfo.VolumeID)
-	publisher.unmountOneAgent(volumeInfo.TargetPath, overlayFSPath)
+	limit := stat.ModTime().Add(volumeCfg.RetryTimeout)
+	log.Info("not first attempt, time remaining before skipping injection", "time", time.Until(limit).String())
 
-	if err = publisher.db.DeleteVolume(ctx, volume.VolumeID); err != nil {
-		return nil, status.Error(codes.Internal, err.Error())
-	}
+	return pub.time.Now().After(limit)
+}
 
-	log.Info("deleted volume info", "ID", volume.VolumeID, "PodUID", volume.PodName, "Version", volume.Version, "TenantUUID", volume.TenantUUID)
+// isArchiveAvailable checks if the LatestAgentBinaryForDynaKube folder exists or not
+func (pub *Publisher) isArchiveAvailable(volumeCfg *csivolumes.VolumeConfig) bool {
```

Comment:
we are not really working with an archive here, but the extracted code module directories

Reply:
[fixed](https://github.com/Dynatrace/dynatrace-operator/pull/4129/commits/3f347a0d75f68e9359fc5389a9ac28e9293c60dc)